File: app/src/main/java/com/inteligenciadigital/netflixremake/MainActivity.java
package com.inteligenciadigital.netflixremake;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.inteligenciadigital.netflixremake.model.Category;
import com.inteligenciadigital.netflixremake.model.Movie;
import com.inteligenciadigital.netflixremake.util.CategoryTask;
import com.inteligenciadigital.netflixremake.util.ImageDownloadTask;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements CategoryTask.CategoryLoader {

	private CategoryAdapter categoryAdapter;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);

		RecyclerView recyclerView = findViewById(R.id.recycler_view_main);

		List<Category> categories = new ArrayList<>();

		this.categoryAdapter = new CategoryAdapter(categories);

		recyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));
		recyclerView.setAdapter(this.categoryAdapter);

		CategoryTask categoryTask = new CategoryTask(this);
		categoryTask.setCategoryLoader(this);
		categoryTask.execute("https://tiagoaguiar.co/api/netflix/home");

	}

	@Override
	public void onResult(List<Category> categories) {
		this.categoryAdapter.setCategories(categories);
		this.categoryAdapter.notifyDataSetChanged();
	}

	private static class MovieHolder extends RecyclerView.ViewHolder {

		private final ImageView imageViewCover;

		public MovieHolder(@NonNull View itemView, final OnClickItemListener onClickItemListener) {
			super(itemView);
			this.imageViewCover = itemView.findViewById(R.id.image_view_cover);
			itemView.setOnClickListener(new View.OnClickListener() {
				@Override
				public void onClick(View view) {
					onClickItemListener.onClick(getAdapterPosition());
				}
			});
		}
	}

	private class MovieAdapter extends RecyclerView.Adapter<MovieHolder>
			implements OnClickItemListener {

		private List<Movie> movies;

		public MovieAdapter(List<Movie> movies) {
			this.movies = movies;
		}

		@NonNull
		@Override
		public MovieHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
			View view = getLayoutInflater().inflate(R.layout.movie_item, parent, false);
			return new MovieHolder(view, this);
		}

		@Override
		public void onBindViewHolder(@NonNull MovieHolder holder, int position) {
			Movie movie = this.movies.get(position);
//			holder.imageViewCover.setImageResource(movie.getCoverUrl());
			new ImageDownloadTask(holder.imageViewCover).execute(movie.getCoverUrl());
		}

		@Override
		public int getItemCount() {
			return this.movies.size();
		}

		@Override
		public void onClick(int position) {
			Intent intent = new Intent(MainActivity.this, MovieActivity.class);
			intent.putExtra("id", this.movies.get(position).getId());
			startActivity(intent);
		}
	}

	private static class CategoryHolder extends RecyclerView.ViewHolder {

		private final TextView textViewTitle;
		private final RecyclerView recyclerViewMovie;

		public CategoryHolder(@NonNull View itemView) {
			super(itemView);
			this.textViewTitle = itemView.findViewById(R.id.text_view_title);
			this.recyclerViewMovie = itemView.findViewById(R.id.recycler_view_movie);
		}
	}

	private class CategoryAdapter extends RecyclerView.Adapter<CategoryHolder> {

		private List<Category> categories;

		public CategoryAdapter(List<Category> categories) {
			this.categories = categories;
		}

		@NonNull
		@Override
		public CategoryHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
			View view = getLayoutInflater().inflate(R.layout.category_item, parent, false);
			return new CategoryHolder(view);
		}

		@Override
		public void onBindViewHolder(@NonNull CategoryHolder holder, int position) {
			Category category = this.categories.get(position);
			holder.textViewTitle.setText(category.getName());
			holder.recyclerViewMovie.setAdapter(new MovieAdapter(category.getMovies()));
			holder.recyclerViewMovie.setLayoutManager(new LinearLayoutManager(getBaseContext(), RecyclerView.HORIZONTAL, false));
		}

		@Override
		public int getItemCount() {
			return this.categories.size();
		}

		public void setCategories(List<Category> categories) {
			this.categories.clear();
			this.categories.addAll(categories);
			this.categories.addAll(categories);
		}
	}

	interface OnClickItemListener {
		void onClick(int position);
	}
}

File: app/src/main/java/com/inteligenciadigital/netflixremake/kotlin/MainActivity.kt
package com.inteligenciadigital.netflixremake.kotlin

import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.inteligenciadigital.netflixremake.R
import com.inteligenciadigital.netflixremake.model.Category
import com.inteligenciadigital.netflixremake.model.Movie
import com.inteligenciadigital.netflixremake.util.CategoryTask
import com.inteligenciadigital.netflixremake.util.ImageDownloadTask
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.category_item.view.*
import kotlinx.android.synthetic.main.movie_item.view.*
import kotlinx.android.synthetic.main.movie_item_similar.view.*
import kotlinx.android.synthetic.main.movie_item_similar.view.image_view_cover


class MainActivity : AppCompatActivity() {

	private lateinit var mainAdapter : MainAdapter

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		this.setContentView(R.layout.activity_main)

		val categories = arrayListOf<Category>()
		this.mainAdapter = MainAdapter(categories)
		recycler_view_main.adapter = this.mainAdapter
		recycler_view_main.layoutManager = LinearLayoutManager(this)

		val categoryTask = CategoryTask(this)
		categoryTask.setCategoryLoader { categories ->
			this.mainAdapter.categories.clear()
			this.mainAdapter.categories.addAll(categories)
			this.mainAdapter.categories.addAll(categories)
			this.mainAdapter.notifyDataSetChanged()
		}
		categoryTask.execute("https://tiagoaguiar.co/api/netflix/home")
	}

	private inner class MainAdapter(val categories: MutableList<Category>) : RecyclerView.Adapter<CategoryHolder>() {
		override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryHolder {
			return CategoryHolder(
					layoutInflater.inflate(
							R.layout.category_item, parent, false)
			)
		}

		override fun onBindViewHolder(holder: CategoryHolder, position: Int) {
			val category = categories[position]
			holder.bind(category)
		}

		override fun getItemCount(): Int = categories.size
	}

	private inner class CategoryHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
		fun bind(category: Category) {
			itemView.text_view_title.text = category.name
			itemView.recycler_view_movie.adapter = MovieAdapter(category.movies)
			itemView.recycler_view_movie.layoutManager = LinearLayoutManager(this@MainActivity, RecyclerView.HORIZONTAL, false)
		}
	}

	private inner class MovieAdapter(val movies: MutableList<Movie>) : RecyclerView.Adapter<MovieHolder>() {
		override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieHolder {
			return MovieHolder(
					layoutInflater.inflate(
							R.layout.movie_item, parent, false)
			)
		}

		override fun onBindViewHolder(holder: MovieHolder, position: Int) {
			val movie = movies[position]
			holder.bind(movie)
		}

		override fun getItemCount(): Int = movies.size
	}

	private class MovieHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
		fun bind(movie: Movie) {
			ImageDownloadTask(itemView.image_view_cover)
					.execute(movie.coverUrl)

			itemView.image_view_cover.setOnClickListener {

			}
		}
	}
}